### PR TITLE
[MemProf] Match function's summary and definition strictly

### PR DESCRIPTION
**Pull request by lifengxiang1025**
Problem description: https://github.com/llvm/llvm-project/pull/81008#issuecomment-1933468520
Solution: https://github.com/llvm/llvm-project/pull/81008#issuecomment-1934192548 (choose plan2)

### CODE REVIEW

**Comment by teresajohnson** [comment]
Thanks for working on this. Since I believe the test case will assert without the added options, I'm thinking it is probably best to simply always do this in a debug compiler. Can we force enable enable-import-metadata when cloning is enabled in the thin link (or pass down a flag to ComputeCrossModuleImport to enable it)? The checking that fails without this patch should probably also go under an `#ifndef NDEBUG` (looks like I missed adding that).

**Comment by lifengxiang1025** [comment]
> Thanks for working on this. Since I believe the test case will assert without the added options, I'm thinking it is probably best to simply always do this in a debug compiler. Can we force enable enable-import-metadata when cloning is enabled in the thin link (or pass down a flag to ComputeCrossModuleImport to enable it)? The checking that fails without this patch should probably also go under an `#ifndef NDEBUG` (looks like I missed adding that).

Sorry, I don't understand fully.
1. I have updated code to match definition and summary default. But I can't understand why we need to distinguish debug compiler and release compiler.
2. I don't understand where to add `#ifndef NDEBUG`.

File: llvm/tools/llvm-link/llvm-link.cpp [hunk 1]
```diff
@@ -134,6 +134,12 @@ static cl::opt<bool> TryUseNewDbgInfoFormat(
     cl::desc("Enable debuginfo iterator positions, if they're built in"),
     cl::init(false));
 
+namespace llvm {
```
Comment (teresajohnson) [inline llvm/tools/llvm-link/llvm-link.cpp:137]: Can we avoid this by moving the def of this flag somewhere common to all uses?

**Comment by teresajohnson** [comment]
> > Thanks for working on this. Since I believe the test case will assert without the added options, I'm thinking it is probably best to simply always do this in a debug compiler. Can we force enable enable-import-metadata when cloning is enabled in the thin link (or pass down a flag to ComputeCrossModuleImport to enable it)? The checking that fails without this patch should probably also go under an `#ifndef NDEBUG` (looks like I missed adding that).
> 
> Sorry, I don't understand fully.
> 
> 1. I have updated code to match definition and summary default. But I can't understand why we need to distinguish debug compiler and release compiler.

Yeah, re-reading my comment and the code this was wrong - I was thinking about the sanity checking we do to ensure the IR and summary match, but we obviously need the correct summary for more than that. Please ignore that part. Doing this always as you have changed the code though does make sense, since there is a correctness issue if we don't find the right summary.

> 2. I don't understand where to add `#ifndef NDEBUG`.

I was looking at the code that does the sanity checking to compare IR and summary, and that whole block can go under the ifndef, but I can do this separately, it is unrelated to your change now.

**Comment by teresajohnson** [comment]
FYI I tried an internal target and I'm getting a failure in applyImport with this change. Let me see if I can identify the reason and will get back to you.

**Comment by teresajohnson** [comment]
> FYI I tried an internal target and I'm getting a failure in applyImport with this change. Let me see if I can identify the reason and will get back to you.

The error I am getting is the assert:
llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:3383 in bool llvm::MemProfContextDisambiguation::applyImport(Module &): SrcModuleMD && "enable-import-metadata is needed to emit thinlto_src_module"

I am passing -enable-memprof-context-disambiguation, however, so I would have thought your change to FunctionImport.cpp should force enable the addition of this metadata. When I explicitly add -enable-import-metadata the assert goes away, so that change isn't having the intended effect. I noticed your test still passes both of these flags, maybe try removing -enable-import-metadata to see if you can reproduce.

**Comment by lifengxiang1025** [comment]
> > FYI I tried an internal target and I'm getting a failure in applyImport with this change. Let me see if I can identify the reason and will get back to you.
> 
> The error I am getting is the assert: llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:3383 in bool llvm::MemProfContextDisambiguation::applyImport(Module &): SrcModuleMD && "enable-import-metadata is needed to emit thinlto_src_module"
> 
> I am passing -enable-memprof-context-disambiguation, however, so I would have thought your change to FunctionImport.cpp should force enable the addition of this metadata. When I explicitly add -enable-import-metadata the assert goes away, so that change isn't having the intended effect. I noticed your test still passes both of these flags, maybe try removing -enable-import-metadata to see if you can reproduce.

Thanks. The reason looks like I omit to emit `thinlto_src_module` to alias. Fix in https://github.com/llvm/llvm-project/pull/83665/commits/caf73e50205b33325065513e8cdbef477716fdc9. Can you try again?

File: llvm/tools/llvm-link/llvm-link.cpp (hunk 1 above)
Comment (lifengxiang1025) [reply llvm/tools/llvm-link/llvm-link.cpp]: https://github.com/llvm/llvm-project/pull/83665/commits/46ddde05cfcb8c141c88d5a7b392441f4e2f3883
Move `EnableMemProfContextDisambiguation` from PassBuilderPipelines.cpp to MemProfContextDisambiguation.cpp. Now there only one definition of `EnableMemProfContextDisambiguation`.

**Review by teresajohnson — APPROVED** [review]
lgtm

**Comment by teresajohnson** [comment]
> > > FYI I tried an internal target and I'm getting a failure in applyImport with this change. Let me see if I can identify the reason and will get back to you.
> > 
> > 
> > The error I am getting is the assert: llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:3383 in bool llvm::MemProfContextDisambiguation::applyImport(Module &): SrcModuleMD && "enable-import-metadata is needed to emit thinlto_src_module"
> > I am passing -enable-memprof-context-disambiguation, however, so I would have thought your change to FunctionImport.cpp should force enable the addition of this metadata. When I explicitly add -enable-import-metadata the assert goes away, so that change isn't having the intended effect. I noticed your test still passes both of these flags, maybe try removing -enable-import-metadata to see if you can reproduce.
> 
> Thanks. The reason looks like I omit to emit `thinlto_src_module` to alias. Fix in [caf73e5](https://github.com/llvm/llvm-project/commit/caf73e50205b33325065513e8cdbef477716fdc9). Can you try again?

Confirming that fixed my issue.

**Pull request merged**